### DISCUSSION TITLE
tutorials/voyage-part2/09-reference-resolvers

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#69] [course feedback]
> **Note: The course/lesson has been updated; this discussion is outdated.**

# tutorials/voyage-part2/09-reference-resolvers

https://www.apollographql.com/tutorials/voyage-part2/09-reference-resolvers

**Comment by davidbuckell** [comment]
In the practice, the Book interface should actually contain the hasPictures property to make it clearer?

interface Book {
  isbn: ID!
  title: String!
  genre: String!
  hasPictures: Boolean
}

**Reply by lizhennessy** [reply]
Hey @davidbuckell ! This is a great question, and an important distinction. In this example, we really want to drive home the fact that our _schema representation_ of an object does not have to match the underlying data object. In this case, the `hasPictures` property that exists on the object we store in the database can be used to differentiate what kind of thing we're working with - a `YoungAdultNovel` or a `PictureBook` - but we aren't actually surfacing it in the schema to be used and consumed by downstream clients. That's certainly something we _could_ do, if there was a need for that data somewhere in the application, but in this case we want to model out how we can access those underlying data properties and make logical decisions based on them. 

I do think we could clarify a bit more where this `hasPictures` property is coming from, so perhaps adding an example object as it might appear in a database would be a good improvement on our end! Please let us know if you have any other questions!

**Comment by rtpa25** [comment]
I think we don't need to define __resolveReference manually anymore, while using apollo server right? {as per voyage I}

**Reply by lizhennessy** [reply]
Hey there @rtpa25 ! We need to define a `__resolveReference` function for an entity whenever the subgraph defines _unique_ fields for it.  Here's a great article in the docs that talks about [the case when we DON'T define the function](https://www.apollographql.com/docs/federation/entities/#referencing-an-entity-without-contributing-fields) - namely, when the subgraph doesn't define any new fields, but merely references in the entity!

In this lesson in Voyage 2, however, we define the `__resolveReference` function in the `accounts` subgraph because it contributes fields to both the `Host` and `Guest` entities. (We moved over the fields `name`, `profilePicture`, and `profileDescription` for `Host`, and `name` and `profilePicture` for `Guest`! Previously these were the responsibility of the `monolith` schema.) Here's the article that [outlines defining an entity across subgraphs](https://www.apollographql.com/docs/federation/entities/), and all of its requirements. I hope that resolves your question - please let us know if anything else

**Comment by ahmadao2214** [comment]
Receiving this response when I test `GetListing` query:
{
  "errors": [
    {
      "message": "Cannot query field \"name\" on type \"Host\".",
      "locations": [
        {
          "line": 4,
          "column": 7
        }
      ],
      "extensions": {
        "code": "GRAPHQL_VALIDATION_FAILED",
        "stacktrace": [
          "GraphQLError: Cannot query field \"name\" on type \"Host\".",
          "    at Object.Field (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/validation/rules/FieldsOnCorrectTypeRule.js:51:13)",
          "    at Object.enter (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/language/visitor.js:301:32)",
          "    at Object.enter (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/utilities/TypeInfo.js:391:27)",
          "    at visit (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/language/visitor.js:197:21)",
          "    at validate (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/validation/validate.js:91:24)",
          "    at processGraphQLRequest (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/@apollo/server/dist/cjs/requestPipeline.js:100:57)",
          "    at runMicrotasks (<anonymous>)",
          "    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
          "    at async internalExecuteOperation (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/@apollo/server/dist/cjs/ApolloServer.js:634:16)",
          "    at async runHttpQuery (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/@apollo/server/dist/cjs/runHttpQuery.js:135:29)"
        ]
      }
    },
    {
      "message": "Cannot query field \"profilePicture\" on type \"Host\".",
      "locations": [
        {
          "line": 5,
          "column": 7
        }
      ],
      "extensions": {
        "code": "GRAPHQL_VALIDATION_FAILED",
        "stacktrace": [
          "GraphQLError: Cannot query field \"profilePicture\" on type \"Host\".",
          "    at Object.Field (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/validation/rules/FieldsOnCorrectTypeRule.js:51:13)",
          "    at Object.enter (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/language/visitor.js:301:32)",
          "    at Object.enter (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/utilities/TypeInfo.js:391:27)",
          "    at visit (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/language/visitor.js:197:21)",
          "    at validate (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/validation/validate.js:91:24)",
          "    at processGraphQLRequest (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/@apollo/server/dist/cjs/requestPipeline.js:100:57)",
          "    at runMicrotasks (<anonymous>)",
          "    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
          "    at async internalExecuteOperation (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/@apollo/server/dist/cjs/ApolloServer.js:634:16)",
          "    at async runHttpQuery (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/@apollo/server/dist/cjs/runHttpQuery.js:135:29)"
        ]
      }
    },
    {
      "message": "Cannot query field \"profileDescription\" on type \"Host\".",
      "locations": [
        {
          "line": 6,
          "column": 7
        }
      ],
      "extensions": {
        "code": "GRAPHQL_VALIDATION_FAILED",
        "stacktrace": [
          "GraphQLError: Cannot query field \"profileDescription\" on type \"Host\".",
          "    at Object.Field (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/validation/rules/FieldsOnCorrectTypeRule.js:51:13)",
          "    at Object.enter (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/language/visitor.js:301:32)",
          "    at Object.enter (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/utilities/TypeInfo.js:391:27)",
          "    at visit (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/language/visitor.js:197:21)",
          "    at validate (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/validation/validate.js:91:24)",
          "    at processGraphQLRequest (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/@apollo/server/dist/cjs/requestPipeline.js:100:57)",
          "    at runMicrotasks (<anonymous>)",
          "    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
          "    at async internalExecuteOperation (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/@apollo/server/dist/cjs/ApolloServer.js:634:16)",
          "    at async runHttpQuery (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/@apollo/server/dist/cjs/runHttpQuery.js:135:29)"
        ]
      }
    },
    {
      "message": "Cannot query field \"name\" on type \"User\".",
      "locations": [
        {
          "line": 12,
          "column": 9
        }
      ],
      "extensions": {
        "code": "GRAPHQL_VALIDATION_FAILED",
        "stacktrace": [
          "GraphQLError: Cannot query field \"name\" on type \"User\".",
          "    at Object.Field (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/validation/rules/FieldsOnCorrectTypeRule.js:51:13)",
          "    at Object.enter (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/language/visitor.js:301:32)",
          "    at Object.enter (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/utilities/TypeInfo.js:391:27)",
          "    at visit (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/language/visitor.js:197:21)",
          "    at validate (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/graphql/validation/validate.js:91:24)",
          "    at processGraphQLRequest (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/@apollo/server/dist/cjs/requestPipeline.js:100:57)",
          "    at runMicrotasks (<anonymous>)",
          "    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
          "    at async internalExecuteOperation (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/@apollo/server/dist/cjs/ApolloServer.js:634:16)",
          "    at async runHttpQuery (/Users/azharahmad/Projects/odyssey-voyage-II-server/monolith/node_modules/@apollo/server/dist/cjs/runHttpQuery.js:135:29)"
        ]
      }
    }
  ]
}


I've tried publishing the final versions of `monolith` and `subgraph-accounts` but it seems to bust when I'm trying to query a field that's not defined in `monolith` for `Host` or `Guest` :(

**Reply by mabuyo** [reply]
Hey there @ahmadao2214 ! Sorry to hear you're having issues. The error `GRAPHQL_VALIDATION_FAILED` points to something wrong in the schema. It seems that the `Host` type doesn't have those fields. Can you check your Schema page in Studio, search for the Host type and confirm that it has the name, profileDescription and profilePicture fields?

**Reply by ahmadao2214** [reply]
I redid the tutorial and I think my hang up was related to publishing? I don't think I respected the `routing-url` port respective to each subgraph (after coming back to it from stepping away). 

Not sure if that was the case given how included `routing-url` when publishing for the first time is mentioned. Unclear what my blocker was, but it's resolved now!

**Reply by ahmadao2214** [reply]
Also noticed that we can eliminate `accountsAPI: new AccountsAPI({ cache })` from `monolith/index.js` given that we successfully moved all references to that `datasource` from `monolith/resolver.js`. This is more a nit

**Reply by ahmadao2214** [reply]
One additional thing that I wish was more clear is how `hostId` or `guestId` is populating when their respective parents have `id` defined as their property within our typedefs. 

I'm not sure if I'm missed something, but it feels like a leap to make that guess that I'm destructing the correct property from the parent object.

**Reply by mabuyo** [reply]
Good question re: `hostId` and `guestId`, we glossed over that explanation in this lesson. Essentially, we need to return an entity representation for a `Host` or `Guest` type: an entity representation in this case is an object that contains the `id` property (the entity's primary key field) representing the host/guest. How do we retrieve that id? 

For a `Host`, the `Listing` type has that information. We can find this by digging into the [ListingsAPI data source](https://github.com/apollographql/odyssey-voyage-II-server/blob/main/services/listings/sequelize/models/listing.js). Since the parent of the `Listing.host` resolver is a `Listing` object, we're assuming that the property is available in the parent. We can console.log the parent to see what object gets returned and confirm that `hostId` exists there. Note that this does not need to be exposed within our _GraphQL schema_. Also note that if that property weren't present in the `parent` object, we might need to make an additional call to our ListingsAPI just to retrieve that information.

For a `Guest`, a similar pattern can be found within the [BookingsAPI data source](https://github.com/apollographql/odyssey-voyage-II-server/blob/main/services/bookings/sequelize/models/booking.js#L11). 

Hope that helps. It's very specific to the implementation of your subgraphs and data sources, but the takeaway here should be that the reference resolver needs to return an entity representation.

**Comment by mshah100** [comment]
I made the changes to index.js, resolver.js etc, but my monolith application fails to start.  Here is the error I am getting:

[nodemon] starting `node index.js`
node:internal/modules/cjs/loader:1147
  throw err;
  ^

Error: Cannot find module 'apollo-server'
Require stack:
- /Users/ms243138/Documents/GraphQL/server/odyssey-voyage-II-server/monolith/resolvers.js
- /Users/ms243138/Documents/GraphQL/server/odyssey-voyage-II-server/monolith/index.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1144:15)
    at Module._load (node:internal/modules/cjs/loader:985:27)
    at Module.require (node:internal/modules/cjs/loader:1235:19)
    at require (node:internal/modules/helpers:176:18)
    at Object.<anonymous> (/Users/ms243138/Documents/GraphQL/server/odyssey-voyage-II-server/monolith/resolvers.js:1:49)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1435:10)
    at Module.load (node:internal/modules/cjs/loader:1207:32)
    at Module._load (node:internal/modules/cjs/loader:1023:12)
    at Module.require (node:internal/modules/cjs/loader:1235:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/ms243138/Documents/GraphQL/server/odyssey-voyage-II-server/monolith/resolvers.js',
    '/Users/ms243138/Documents/GraphQL/server/odyssey-voyage-II-server/monolith/index.js'

  ]
}


index.js in monolith
const { ApolloServer } = require('@apollo/server');
const { startStandaloneServer } = require('@apollo/server/standalone');

const { readFileSync } = require('fs');
const axios = require('axios');
const gql = require('graphql-tag');

const { AuthenticationError } = require('./utils/errors');

const typeDefs = gql(readFileSync('./schema.graphql', { encoding: 'utf-8' }));
const resolvers = require('./resolvers');

const BookingsDataSource = require('./datasources/bookings');
const ReviewsDataSource = require('./datasources/reviews');
const ListingsAPI = require('./datasources/listings');
const AccountsAPI = require('./datasources/accounts');
const PaymentsAPI = require('./datasources/payments');

const { buildSubgraphSchema } = require("@apollo/subgraph");

async function startApolloServer() {
  const server = new ApolloServer({
    schema: buildSubgraphSchema({
      typeDefs,
      resolvers,
    }),
  });

  const port = 4001;

  try {
    const { url } = await startStandaloneServer(server, {
      context: async ({ req }) => {
        const token = req.headers.authorization || '';
        const userId = token.split(' ')[1]; // get the user name after 'Bearer '

        let userInfo = {};
        if (userId) {
          const { data } = await axios
            .get(`http://localhost:4011/login/${userId}`)
            .catch((error) => {
              throw AuthenticationError();
            });

          userInfo = { userId: data.id, userRole: data.role };
        }

        const { cache } = server;

        return {
          ...userInfo,
          dataSources: {
            bookingsDb: new BookingsDataSource(),
            reviewsDb: new ReviewsDataSource(),
            listingsAPI: new ListingsAPI({ cache }),
            accountsAPI: new AccountsAPI({ cache }),
            paymentsAPI: new PaymentsAPI({ cache }),
          },
        };
      },
      listen: {
        port,
      },
    });

    console.log(`🚀  Server ready at ${url}`);
  } catch (err) {
    console.error(err);
  }
}

startApolloServer();


resolver.js file:
const { AuthenticationError, ForbiddenError } = require('apollo-server');
const authErrMessage = '*** you must be logged in ***';

const resolvers = {
  Query: {
    user: async (_, { id }, { dataSources }) => {
      const user = await dataSources.accountsAPI.getUser(id);
      if (!user) {
        throw new Error('No user found for this Id');
      }
      return user;
    },
    me: async (_, __, { dataSources, userId }) => {
      if (!userId) throw new AuthenticationError(authErrMessage);
      const user = await dataSources.accountsAPI.getUser(userId);
      return user;
    },
    searchListings: async (_, { criteria }, { dataSources }) => {
      const { numOfBeds, checkInDate, checkOutDate, page, limit, sortBy } = criteria;
      const listings = await dataSources.listingsAPI.getListings({ numOfBeds, page, limit, sortBy });

      // check availability for each listing
      const listingAvailability = await Promise.all(
        listings.map((listing) =>
          dataSources.bookingsDb.isListingAvailable({ listingId: listing.id, checkInDate, checkOutDate })
        )
      );

      // filter listings data based on availability
      const availableListings = listings.filter((listing, index) => listingAvailability[index]);

      return availableListings;
    },
    hostListings: async (_, __, { dataSources, userId, userRole }) => {
      if (!userId) throw new AuthenticationError(authErrMessage);

      if (userRole === 'Host') {
        return dataSources.listingsAPI.getListingsForUser(userId);
      } else {
        throw new ForbiddenError('Only hosts have access to listings.');
      }
    },
    listing: (_, { id }, { dataSources }) => {
      return dataSources.listingsAPI.getListing(id);
    },
    featuredListings: (_, __, { dataSources }) => {
      const limit = 3;
      return dataSources.listingsAPI.getFeaturedListings(limit);
    },
    listingAmenities: (_, __, { dataSources }) => {
      return dataSources.listingsAPI.getAllAmenities();
    },
    guestBookings: async (_, __, { dataSources, userId, userRole }) => {
      if (!userId) throw new AuthenticationError(authErrMessage);

      if (userRole === 'Guest') {
        const bookings = await dataSources.bookingsDb.getBookingsForUser(userId);
        return bookings;
      } else {
        throw new ForbiddenError('Only guests have access to trips');
      }
    },
    upcomingGuestBookings: async (_, __, { dataSources, userId, userRole }) => {
      if (!userId) throw new AuthenticationError(authErrMessage);

      if (userRole === 'Guest') {
        const bookings = await dataSources.bookingsDb.getBookingsForUser(userId, 'UPCOMING');
        return bookings;
      } else {
        throw new ForbiddenError('Only guests have access to trips');
      }
    },
    pastGuestBookings: async (_, __, { dataSources, userId, userRole }) => {
      if (!userId) throw new AuthenticationError(authErrMessage);

      if (userRole === 'Guest') {
        const bookings = await dataSources.bookingsDb.getBookingsForUser(userId, 'COMPLETED');
        return bookings;
      } else {
        throw new ForbiddenError('Only guests have access to trips');
      }
    },
    bookingsForListing: async (_, { listingId, status }, { dataSources, userId, userRole }) => {
      if (!userId) throw new AuthenticationError(authErrMessage);

      if (userRole === 'Host') {
        // need to check if listing belongs to host
        const listings = await dataSources.listingsAPI.getListingsForUser(userId);
        if (listings.find((listing) => listing.id === listingId)) {
          const bookings = (await dataSources.bookingsDb.getBookingsForListing(listingId, status)) || [];
          return bookings;
        } else {
          throw new Error('Listing does not belong to host');
        }
      } else {
        throw new ForbiddenError('Only hosts have access to listing bookings');
      }
    },
  },
  Mutation: {
    updateProfile: async (_, { updateProfileInput }, { dataSources, userId }) => {
      if (!userId) throw new AuthenticationError(authErrMessage);
      try {
        const updatedUser = await dataSources.accountsAPI.updateUser({ userId, userInfo: updateProfileInput });
        return {
          code: 200,
          success: true,
          message: 'Profile successfully updated!',
          user: updatedUser,
        };
      } catch (err) {
        return {
          code: 400,
          success: false,
          message: err.message,
        };
      }
    },
    createBooking: async (_, { createBookingInput }, { dataSources, userId }) => {
      if (!userId) throw new AuthenticationError(authErrMessage);

      const { listingId, checkInDate, checkOutDate } = createBookingInput;
      const { totalCost } = await dataSources.listingsAPI.getTotalCost({ id: listingId, checkInDate, checkOutDate });

      try {
        await dataSources.paymentsAPI.subtractFunds({ userId, amount: totalCost });
      } catch (e) {
        return {
          code: 400,
          success: false,
          message: 'We couldn’t complete your request because your funds are insufficient.',
        };
      }

      try {
        const booking = await dataSources.bookingsDb.createBooking({
          listingId,
          checkInDate,
          checkOutDate,
          totalCost,
          guestId: userId,
        });

        return {
          code: 200,
          success: true,
          message: 'Successfully booked!',
          booking,
        };
      } catch (err) {
        return {
          code: 400,
          success: false,
          message: err.message,
        };
      }
    },
    createListing: async (_, { listing }, { dataSources, userId, userRole }) => {
      if (!userId) throw new AuthenticationError(authErrMessage);

      const { title, description, photoThumbnail, numOfBeds, costPerNight, locationType, amenities } = listing;

      if (userRole === 'Host') {
        try {
          const newListing = await dataSources.listingsAPI.createListing({
            title,
            description,
            photoThumbnail,
            numOfBeds,
            costPerNight,
            hostId: userId,
            locationType,
            amenities,
          });

          return {
            code: 200,
            success: true,
            message: 'Listing successfully created!',
            listing: newListing,
          };
        } catch (err) {
          return {
            code: 400,
            success: false,
            message: err.message,
          };
        }
      } else {
        return {
          code: 400,
          success: false,
          message: 'Only hosts can create new listings',
        };
      }
    },
    updateListing: async (_, { listingId, listing }, { dataSources, userId }) => {
      if (!userId) throw new AuthenticationError(authErrMessage);

      try {
        const updatedListing = await dataSources.listingsAPI.updateListing({ listingId, listing });

        return {
          code: 200,
          success: true,
          message: 'Listing successfully updated!',
          listing: updatedListing,
        };
      } catch (err) {
        return {
          code: 400,
          success: false,
          message: err.message,
        };
      }
    },
    submitGuestReview: async (_, { bookingId, guestReview }, { dataSources, userId }) => {
      if (!userId) throw new AuthenticationError(authErrMessage);

      const { rating, text } = guestReview;
      const guestId = await dataSources.bookingsDb.getGuestIdForBooking(bookingId);

      const createdReview = await dataSources.reviewsDb.createReviewForGuest({
        bookingId,
        guestId,
        authorId: userId,
        text,
        rating,
      });
      return {
        code: 200,
        success: true,
        message: 'Successfully submitted review for guest',
        guestReview: createdReview,
      };
    },
    submitHostAndLocationReviews: async (_, { bookingId, hostReview, locationReview }, { dataSources, userId }) => {
      if (!userId) throw new AuthenticationError(authErrMessage);

      const listingId = await dataSources.bookingsDb.getListingIdForBooking(bookingId);
      const createdLocationReview = await dataSources.reviewsDb.createReviewForListing({
        bookingId,
        listingId,
        authorId: userId,
        text: locationReview.text,
        rating: locationReview.rating,
      });

      const { hostId } = await dataSources.listingsAPI.getListing(listingId);
      const createdHostReview = await dataSources.reviewsDb.createReviewForHost({
        bookingId,
        hostId,
        authorId: userId,
        text: hostReview.text,
        rating: hostReview.rating,
      });

      return {
        code: 200,
        success: true,
        message: 'Successfully submitted review for host and location',
        hostReview: createdHostReview,
        locationReview: createdLocationReview,
      };
    },
    addFundsToWallet: async (_, { amount }, { dataSources, userId }) => {
      if (!userId) throw new AuthenticationError(authErrMessage);
      try {
        const updatedWallet = await dataSources.paymentsAPI.addFunds({ userId, amount });
        return {
          code: 200,
          success: true,
          message: 'Successfully added funds to wallet',
          amount: updatedWallet.amount,
        };
      } catch (err) {
        return {
          code: 400,
          success: false,
          message: err.message,
        };
      }
    },
  },
  Host: {
    overallRating: ({ id }, _, { dataSources }) => {
      return dataSources.reviewsDb.getOverallRatingForHost(id);
    },
  },
  Guest: {
    funds: async (_, __, { dataSources, userId }) => {
      const { amount } = await dataSources.paymentsAPI.getUserWalletAmount(userId);
      return amount;
    },
  },
  Listing: {
    host: ({ hostId }) => {
      return { id: hostId };
    },
    overallRating: ({ id }, _, { dataSources }) => {
      return dataSources.reviewsDb.getOverallRatingForListing(id);
    },
    reviews: ({ id }, _, { dataSources }) => {
      return dataSources.reviewsDb.getReviewsForListing(id);
    },
    totalCost: async ({ id }, { checkInDate, checkOutDate }, { dataSources }) => {
      const { totalCost } = await dataSources.listingsAPI.getTotalCost({ id, checkInDate, checkOutDate });
      return totalCost;
    },
    currentlyBookedDates: ({ id }, _, { dataSources }) => {
      return dataSources.bookingsDb.getCurrentlyBookedDateRangesForListing(id);
    },
    bookings: ({ id }, _, { dataSources }) => {
      return dataSources.bookingsDb.getBookingsForListing(id);
    },
    numberOfUpcomingBookings: async ({ id }, _, { dataSources }) => {
      const bookings = (await dataSources.bookingsDb.getBookingsForListing(id, 'UPCOMING')) || [];
      return bookings.length;
    },
  },
  Booking: {
    listing: ({ listingId }, _, { dataSources }) => {
      return dataSources.listingsAPI.getListing(listingId);
    },
    checkInDate: ({ checkInDate }, _, { dataSources }) => {
      return dataSources.bookingsDb.getHumanReadableDate(checkInDate);
    },
    checkOutDate: ({ checkOutDate }, _, { dataSources }) => {
      return dataSources.bookingsDb.getHumanReadableDate(checkOutDate);
    },
    guest: ({ guestId }) => {
      return { id: guestId };
    },
    totalPrice: async ({ listingId, checkInDate, checkOutDate }, _, { dataSources }) => {
      const { totalCost } = await dataSources.listingsAPI.getTotalCost({ id: listingId, checkInDate, checkOutDate });
      return totalCost;
    },
    guestReview: ({ id }, _, { dataSources }) => {
      return dataSources.reviewsDb.getReviewForBooking('GUEST', id);
    },
    hostReview: ({ id }, _, { dataSources }) => {
      return dataSources.reviewsDb.getReviewForBooking('HOST', id);
    },
    locationReview: ({ id }, _, { dataSources }) => {
      return dataSources.reviewsDb.getReviewForBooking('LISTING', id);
    },
  },
  Review: {
    author: (review) => {
      let role = '';
      if (review.targetType === 'LISTING' || review.targetType === 'HOST') {
        role = 'Guest';
      } else { 
        role = 'Host';
      }
      return { __typename: role, id: review.authorId };
    },
  },
  AmenityCategory: {
    ACCOMMODATION_DETAILS: 'Accommodation Details',
    SPACE_SURVIVAL: 'Space Survival',
    OUTDOORS: 'Outdoors',
  },
};

module.exports = resolvers;

Any inputs on why I am getting this error will be appreciated.

**Reply by mshah100** [reply]
I fixed this by including the buildSubGraphSchema call.

**Reply by mabuyo** [reply]
Glad you resolved the error!

**Comment by neoVoda** [comment]
This is the error i get: {
  "data": {
    "listing": null
  },
  "errors": [
    {
      "message": "HTTP fetch failed from 'accounts': HTTP fetch failed from 'accounts': error trying to connect: tcp connect error: No connection could be made because the target machine actively refused it. (os error 10061)",
      "path": [
        "listing",
        "host"
      ],
      "extensions": {
        "code": "SUBREQUEST_HTTP_ERROR",
        "service": "accounts",
        "reason": "HTTP fetch failed from 'accounts': error trying to connect: tcp connect error: No connection could be made because the target machine actively refused it. (os error 10061)"
      }
    },
    {
      "message": "HTTP fetch failed from 'accounts': HTTP fetch failed from 'accounts': error trying to connect: tcp connect error: No connection could be made because the target machine actively refused it. (os error 10061)",
      "path": [
        "listing",
        "reviews",
        "@",
        "author"
      ],
      "extensions": {
        "code": "SUBREQUEST_HTTP_ERROR",
        "service": "accounts",
        "reason": "HTTP fetch failed from 'accounts': error trying to connect: tcp connect error: No connection could be made because the target machine actively refused it. (os error 10061)"
      }
    }
  ],
  "extensions": {
    "valueCompletion": [
      {
        "message": "Cannot return null for non-nullable field Host.name",
        "path": [
          "listing",
          "host"
        ]
      },
      {
        "message": "Cannot return null for non-nullable field Listing.host",
        "path": [
          "listing",
          "host"
        ]
      }
    ]
  }
}

**Reply by mabuyo** [reply]
This error looks like there's something wrong with the `accounts` subgraph. Can you confirm it's running successfully?

**Comment by rickyssj2** [comment]
Why does my code not work?

// Here's the solution (your code is commented out after):
const resolvers = {
  LibraryMember: {
    faveBook: (book) => {
      const type = book.hasPictures ? 'PictureBook' : 'YoungAdultNovel';
      return {__typename: type, isbn: book.isbn}
    }
  }
}
/* your code:
const resolvers = {
  LibraryMember: {
    faveBook: (book) => {
      let type ='';
      if (book.hasPictures) {
        type = 'PictureBook';
      } else {
        type = 'YoungAdultNovel';
      }

      return {__typeName: type, isbn: book.isbn};
    }
  }
}
*/

**Reply by lizhennessy** [reply]
Hi @rickyssj2 ! It looks like the problem is specifically with the `__typeName` property in the object your resolver returns. This property _is_ case-sensitive, and should be written as `__typename` instead; no camel case. :) I hope this helps!
